pkg/systemlogmonitor/counts_buffer.go

```diff
@@ -0,0 +1,87 @@
+/*
+Copyright 2023 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package systemlogmonitor
+
+import (
+	"fmt"
+	"github.com/golang/glog"
+	systemlogtypes "k8s.io/node-problem-detector/pkg/systemlogmonitor/types"
+	"time"
+)
+
+const (
+	MaxCountRuleMatched = 128
+	MaxTimeExpired      = time.Hour * 168
+)
+
+type CountRingBuffer struct {
+	data         []time.Time
+	index        int
+	length       int
+	expirePeriod time.Duration
+}
+
+func NewCountRingBuffer(thd int, ep time.Duration) *CountRingBuffer {
+	return &CountRingBuffer{
+		data:         make([]time.Time, thd),
+		index:        0,
+		length:       thd,
+		expirePeriod: ep,
+	}
+}
+
+func (crb *CountRingBuffer) IsMatchthreshold() bool {
+	tempIndex := crb.index
+	crb.index = (crb.index + 1) % crb.length
+    if crb.data[crb.index].IsZero() {
+		crb.data[tempIndex] = time.Now()
```

Comment:
Also, why are we using `Now()`. Shouldn't we use the time when the log entry appeared?

Reply:
I'm not sure if the implementation actually depends on "now". But if it does, please consider using a clock instead, that way we can mock it in tests. See https://github.com/kubernetes/node-problem-detector/blob/fd51f17ec1f7f85086556c97b0d5efb40ead0fd6/pkg/systemlogmonitor/logwatchers/filelog/log_watcher.go#L65 for an example.